### PR TITLE
fuzzTests optimization

### PR DESCRIPTION
**Pull request by winsvega**
add boost test
pass random code options as a reference

### CODE REVIEW

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 1]
```diff
@@ -85,17 +79,18 @@ int fillRandomTest(dev::test::TestSuite const& _testSuite, std::string const& _t
...
+		std::cout << json_spirit::write_string(v, true);
 
-	std::cout << json_spirit::write_string(v, true);
-	return 0;
+	return json_spirit::write_string(v, true);
 }
 
 /// Parse Test string replacing keywords to fuzzed values
-void dev::test::RandomCode::parseTestWithTypes(std::string& _test, std::map<std::string, std::string> const& _varMap)
+void dev::test::RandomCode::parseTestWithTypes(std::string& _test, std::map<std::string, std::string> const& _varMap, RandomCodeOptions& _options)
```
Comment (pirapira) [inline test/tools/fuzzTesting/createRandomTest.cpp]: Can this be `RandomCodeOptions const&`?

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 2]
```diff
@@ -85,17 +79,18 @@ int fillRandomTest(dev::test::TestSuite const& _testSuite, std::string const& _t
...
 {
-	dev::test::RandomCodeOptions options; //use default options
+	dev::test::RandomCodeOptions& options = _options;
```
Comment (pirapira) [inline test/tools/fuzzTesting/createRandomTest.cpp]: You can remove this line, and use `_options` below.

File: test/tools/fuzzTesting/createRandomTest.cpp (hunk 1 above)
Comment (winsvega) [reply test/tools/fuzzTesting/createRandomTest.cpp]: I guess no. if it use random generators it can't be const. and we want it to use both static (from input) and random code generation.

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 3]
```diff
@@ -253,7 +254,7 @@ std::string RandomCode::rndByteSequence(int _length, SizeStrictness _sizeType)
 }
 
 //generate smart random code
-std::string RandomCode::generate(int _maxOpNumber, RandomCodeOptions _options)
+std::string RandomCode::generate(int _maxOpNumber, RandomCodeOptions& _options)
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: Can this be `RandomCodeOptions const&`?

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 4]
```diff
@@ -564,9 +563,24 @@ BOOST_AUTO_TEST_CASE(rndCode)
...
-		std::string code = test::RandomCode::generate(1000);
-		cnote << code;
+		test::RandomCodeOptions options;
+		std::string code = test::RandomCode::generate(1000, options);
+		BOOST_REQUIRE(!code.empty());
+	}
+	catch(...)
+	{
+		BOOST_ERROR("Exception thrown when generating random code!");
+	}
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: Can't there be a way to show what kind of exception it was.  At least show `e.what()`.

File: test/tools/fuzzTesting/createRandomTest.cpp (hunk 1 above)
Comment (pirapira) [reply test/tools/fuzzTesting/createRandomTest.cpp]: Which field has to be modified?

File: test/tools/fuzzTesting/createRandomTest.cpp [hunk 5]
```diff
@@ -29,17 +29,8 @@
...
-extern std::string const c_testExampleVMTest;
-extern std::string const c_testExampleBlockchainTest;
-extern std::string const c_testExampleRLPTest;
-
-//Main Test functinos
-int fillRandomTest(dev::test::TestSuite const& _testSuite, std::string const& _testString);
-
 namespace dev { namespace test {
+
 int createRandomTest()
```
Comment (pirapira) [inline test/tools/fuzzTesting/createRandomTest.cpp]: If the return value is either `true` or `false`, make it a `bool`.

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 6]
```diff
@@ -341,7 +340,7 @@ std::string RandomCode::getPushCode(int _value)
 	return getPushCode(hexString);
 }
 
-std::string RandomCode::fillArguments(eth::Instruction _opcode, RandomCodeOptions const& _options)
+std::string RandomCode::fillArguments(eth::Instruction _opcode, RandomCodeOptions& _options)
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: Why did you need to remove `const`?

File: test/tools/fuzzTesting/fuzzHelper.cpp (hunk 3 above)
Comment (winsvega) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: hmm. although if I move opCodeProbability out of options I guess it might become const

File: test/tools/fuzzTesting/fuzzHelper.h [hunk 7]
```diff
@@ -108,7 +128,7 @@ class RandomCode
 	static int randomPercent() { refreshSeed(); return percentDist(gen); }
 
 private:
-	static std::string fillArguments(dev::eth::Instruction _opcode, RandomCodeOptions const& _options);
+	static std::string fillArguments(dev::eth::Instruction _opcode, RandomCodeOptions& _options);
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.h]: Why did you need to remove `const`?

File: test/tools/fuzzTesting/createRandomTest.cpp (hunk 1 above)
Comment (pirapira) [reply test/tools/fuzzTesting/createRandomTest.cpp]: Why do you think it can't be const when it uses random generators?

File: test/tools/fuzzTesting/fuzzHelper.cpp (hunk 3 above)
Comment (pirapira) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: Which line changes `opCodeProbability`?

File: test/tools/libtesteth/TestHelper.h [hunk 8]
```diff
@@ -112,7 +112,7 @@ json_spirit::mObject fillJsonWithState(eth::State const& _state, eth::AccountMas
 json_spirit::mObject fillJsonWithTransaction(eth::Transaction const& _txn);
 
 //Fill Test Functions
-int createRandomTest();	//returns 0 if succeed, 1 if there was an error;
+bool createRandomTest();	//returns 0 if succeed, 1 if there was an error;
```
Comment (pirapira) [inline test/tools/libtesteth/TestHelper.h]: Update the comment, and everywhere that uses the return value.

Comment (winsvega) [reply test/tools/libtesteth/TestHelper.h]: what should be updated everywhere that uses the return value?

File: test/tools/fuzzTesting/fuzzHelper.cpp [hunk 9]
```diff
@@ -550,12 +547,19 @@ Address RandomCodeOptions::getRandomAddress(AddressType _type) const
 	}
 }
 
-void RandomCodeOptions::setWeights()
+int RandomCode::weightedOpcode(std::vector<int>& _weights)
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.cpp]: Can this function receive `opCodeProbability` instead of `_weights`?  Then, you don't need the static variable.  Static variables are like global variables, so let's try not to introduce them.

Comment (pirapira) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: Alternatively, maybe you can make `opCodeProbability` a local variable in this function.  That seems cleaner.

File: test/tools/fuzzTesting/fuzzHelper.h [hunk 10]
```diff
@@ -121,6 +138,7 @@ class RandomCode
 	static IntDistrib opLengDist;			///< 1..32  byte string
 	static IntDistrib opMemrDist;			///< 1..10MB  byte string
 	static IntDistrib uniIntDist;			///< 0..0x7fffffff
+	static DescreteDistrib opCodeProbability; ///< An Opcode distribution with weights
```
Comment (pirapira) [inline test/tools/fuzzTesting/fuzzHelper.h]: Let's try not to introduce a new static variable.

File: test/tools/libtesteth/TestHelper.h (hunk 8 above)
Comment (pirapira) [reply test/tools/libtesteth/TestHelper.h]: This function is used somewhere, no?

File: test/tools/fuzzTesting/fuzzHelper.cpp (hunk 9 above)
Comment (winsvega) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: from where should it receve opCodeProbability ?

File: test/tools/fuzzTesting/fuzzHelper.h (hunk 10 above)
Comment (winsvega) [reply test/tools/fuzzTesting/fuzzHelper.h]: would it be better if I declare Random code as single tone?

File: test/tools/fuzzTesting/fuzzHelper.cpp (hunk 9 above)
Comment (winsvega) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: the idea is that random generators are managed by random code class.

File: test/tools/libtesteth/TestHelper.h (hunk 8 above)
Comment (pirapira) [reply test/tools/libtesteth/TestHelper.h]: Before, `0` was success and `1` was failure.  Now, `false` is failure and `true` is success.

Remember in C++, `if (false)` and `if(0)` is very similar.  So, something is probably broken.

File: test/tools/fuzzTesting/fuzzHelper.cpp (hunk 9 above)
Comment (pirapira) [reply test/tools/fuzzTesting/fuzzHelper.cpp]: The caller.  Read my second comment as well, and choose one of my two suggestions (or something else of course).

File: test/tools/fuzzTesting/fuzzHelper.h (hunk 10 above)
Comment (pirapira) [reply test/tools/fuzzTesting/fuzzHelper.h]: tone?

Comment (winsvega) [reply test/tools/fuzzTesting/fuzzHelper.h]: so like all of those fields would be private fields of a class.
and the only static instance would be an instance of that class

File: test/tools/libtesteth/TestHelper.h (hunk 8 above)
Comment (winsvega) [reply test/tools/libtesteth/TestHelper.h]: this one is only used in create random code. I always check that --createRandomCode option is working

Comment (pirapira) [reply test/tools/libtesteth/TestHelper.h]: Does it check the return value of `createRandomTest()`?

File: test/tools/fuzzTesting/fuzzHelper.h (hunk 10 above)
Comment (pirapira) [reply test/tools/fuzzTesting/fuzzHelper.h]: First, try to remove this line.

File: test/tools/libtesteth/TestHelper.h (hunk 8 above)
Comment (winsvega) [reply test/tools/libtesteth/TestHelper.h]: hm. that changes are in my local repo. can't link yet.

File: test/tools/libtesteth/boostTest.cpp [hunk 11]
```diff
@@ -90,37 +86,32 @@ void setDefaultOrCLocale()
...
-				if (arg == "--checktest")
 				{
-					argc = i + 1;
+					argv[i+1] = (char*)dynamicTestSuiteName.c_str();
 					break;
 				}
 			}
 
 			//add random tests suite
-			// FIXME:
```
Comment (pirapira) [inline test/tools/libtesteth/boostTest.cpp]: Do you remember what was the idea behind this `FIXME`?

File: test/tools/libtesteth/boostTest.cpp [hunk 12]
```diff
@@ -90,37 +86,32 @@ void setDefaultOrCLocale()
...
 				//replace test suite to random tests
+				std::string arg = std::string{argv[i]};
 				if (arg == "-t" && i+1 < argc)
-					argv[i+1] = (char*)std::string("RandomTestCreationSuite").c_str();
-
-				//don't pass long raw test input to boost
-				if (arg == "--checktest")
 				{
-					argc = i + 1;
+					argv[i+1] = (char*)dynamicTestSuiteName.c_str();
```
Comment (pirapira) [inline test/tools/libtesteth/boostTest.cpp]: `i+1` should be `i + 1` with spaces.

Comment (pirapira) [inline test/tools/libtesteth/boostTest.cpp]: Before doing this assignment, make sure `argc > i + 1` holds.

Comment (winsvega) [reply test/tools/libtesteth/boostTest.cpp]: this is done one line up

File: test/tools/libtesteth/boostTest.cpp (hunk 11 above)
Comment (winsvega) [reply test/tools/libtesteth/boostTest.cpp]: No. I don't think that is was my comment either.

File: test/tools/libtesteth/boostTest.cpp (hunk 12 above)
Comment (pirapira) [reply test/tools/libtesteth/boostTest.cpp]: OK.  The next question: why is it necessary to edit the user input?

File: test/tools/libtesteth/boostTest.cpp (hunk 11 above)
Comment (pirapira) [reply test/tools/libtesteth/boostTest.cpp]: Then leave it.

Comment (pirapira) [reply test/tools/libtesteth/boostTest.cpp]: Please recover the `FIXME`.

**Comment by winsvega** [comment]
why? does it says anything about whats needed to be fixed? what is the point of returning it?

File: test/tools/libtesteth/boostTest.cpp (hunk 12 above)
Comment (winsvega) [reply test/tools/libtesteth/boostTest.cpp]: this is a trick to overwrite boost option -t GeneralStateTests     with   -t randomTestCreationSuite which is dynamicly generated.

Comment (winsvega) [reply test/tools/libtesteth/boostTest.cpp]: so you type  -t GeneralStateTests -- --createRandomTest  
but instead it runs boost with -t randomTestCreationSuite 
which then reads what particular test to generate. in this case it is GeneralStateTest

**Comment by pirapira** [comment]
I can read it like "random test suite should not be added here, this needs to be fixed.".

File: test/tools/libtesteth/boostTest.cpp (hunk 12 above)
Comment (pirapira) [reply test/tools/libtesteth/boostTest.cpp]: Will you add a comment explaining these?
